File: python/1793.maximum-score-of-a-good-subarray/solution.py
# Created by Jones at 2023/10/22 14:25
# leetgo: dev
# https://leetcode.cn/problems/maximum-score-of-a-good-subarray/

from typing import *
from leetgo_py import *

from bisect import bisect_left, bisect_right
from collections import Counter, defaultdict, deque
from copy import deepcopy
from functools import cache, cmp_to_key, lru_cache, reduce
from heapq import heapify, heappop, heappush, heappushpop, heapreplace
from itertools import accumulate, chain, count, pairwise, zip_longest
from math import ceil, comb, floor, gcd, inf, isqrt, log2, perm, sqrt
from operator import xor
from string import ascii_lowercase
from typing import List, Optional

# @lc code=begin
# from sortedcontainers import SortedList


class Solution:
    def maximumScore(self, nums: List[int], k: int) -> int:
        n = len(nums)

        left = [0] * n
        right = [n - 1] * n

        # find the first j of a[j] > a[i]

        st = []

        for i, x in enumerate(nums):
            while st and x < nums[st[-1]]:
                right[st.pop()] = i - 1
            if st:
                left[i] = st[-1] + 1
            st.append(i)

        res = 0
        for i in range(n):
            l, r = left[i], right[i]
            if l <= k <= r:
                res = max(res, (r - l + 1) * nums[i])
        return res


# @lc code=end

if __name__ == "__main__":
    nums: List[int] = deserialize("List[int]", read_line())
    k: int = deserialize("int", read_line())
    ans = Solution().maximumScore(nums, k)

    print("\noutput:", serialize(ans))

File: rust/src/1793.maximum-score-of-a-good-subarray/solution.rs
// Created by Jones at 2024/03/19 15:33
// leetgo: 1.4.2
// https://leetcode.cn/problems/maximum-score-of-a-good-subarray/

use anyhow::Result;
use leetgo_rs::*;

struct Solution;

// @lc code=begin
use std::cmp::Reverse;
use std::mem::swap;

impl Solution {
    pub fn maximum_score(nums: Vec<i32>, k: i32) -> i32 {
        let n = nums.len();
        let k = k as usize;

        // find the range of nums[i] is minimal
        let mut left = vec![0; n];
        let mut right = vec![n - 1; n];

        let mut st = vec![];
        for (i, &x) in nums.iter().enumerate() {
            while let Some(j) = st.pop() {
                if x < nums[j] {
                    right[j] = i - 1;
                } else {
                    st.push(j);
                    break;
                }
            }
            if let Some(j) = st.last() {
                left[i] = *j + 1;
            }
            st.push(i)
        }

        let mut res = 0;
        for i in 0..n {
            let (l, r) = (left[i], right[i]);
            if l <= k && k <= r {
                res = res.max(nums[i] * (r - l + 1) as i32)
            }
        }
        res
    }
}

// @lc code=end

fn main() -> Result<()> {
    let nums: Vec<i32> = deserialize(&read_line()?)?;
    let k: i32 = deserialize(&read_line()?)?;
    let ans: i32 = Solution::maximum_score(nums, k).into();

    println!("\noutput: {}", serialize(ans)?);
    Ok(())
}
